src/shogun/clustering/KMeansBase.cpp

```diff
@@ -0,0 +1,405 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 1999-2008 Gunnar Raetsch
+ * Written (W) 2007-2009 Soeren Sonnenburg
+ * Copyright (C) 1999-2009 Fraunhofer Institute FIRST and Max-Planck-Society
+ */
+
+#include <shogun/clustering/KMeansBase.h>
+#include <shogun/distance/Distance.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/base/Parallel.h>
+
+#ifdef HAVE_PTHREAD
+#include <pthread.h>
+#endif
+
+#define MUSRECALC
+
+using namespace shogun;
+
+CKMeansBase::CKMeansBase()
+: CDistanceMachine()
+{
+	init();
+}
+
+CKMeansBase::CKMeansBase(int32_t k_, CDistance* d, bool use_kmpp)
+: CDistanceMachine()
+{	
+	init();
+	k=k_;
+	set_distance(d);
+	use_kmeanspp=use_kmpp;
+}
+
+CKMeansBase::CKMeansBase(int32_t k_i, CDistance* d_i, SGMatrix<float64_t> centers_i)
+: CDistanceMachine()
+{
+	init();
+	k = k_i;
+	set_distance(d_i);
+	set_initial_centers(centers_i);
+}
+
+CKMeansBase::~CKMeansBase()
+{
+}	
+
+void CKMeansBase::set_initial_centers(SGMatrix<float64_t> centers)
+{	
+	
+	dimensions = ((CDenseFeatures<float64_t>*) distance->get_lhs())->get_num_features();
+
+	REQUIRE(centers.num_cols == k,
+			"Expected %d initial cluster centers, got %d", k, centers.num_cols);
+
+	REQUIRE(centers.num_rows == dimensions,
+			"Expected %d dimensionional cluster centers, got %d", dimensions, centers.num_rows);
+
+	m_mus_initial = centers;
+}
+
+void CKMeansBase::set_random_centers(SGVector<float64_t> weights_set, SGVector<int32_t> cl_list, int32_t XSize)
+{
+	CDenseFeatures<float64_t>* lhs=
+		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
+
+	for (int32_t i=0; i<XSize; i++)
+	{
+		const int32_t Cl=CMath::random(0, k-1);
+		weights_set[Cl]+=1.0;
+		cl_list[i]=Cl;
+
+		int32_t vlen=0;
+		bool vfree=false;
+		float64_t* vec=lhs->get_feature_vector(i, vlen, vfree);
+
+		for (int32_t j=0; j<dimensions; j++)
+			m_mus(j, Cl) += vec[j];
+
+		lhs->free_feature_vector(vec, i, vfree);
+	}
+
+	SG_UNREF(lhs);
+
+	for (int32_t i=0; i<k; i++)
+	{
+		if (weights_set[i]!=0.0)
+		{
+			for (int32_t j=0; j<dimensions; j++)
+				m_mus(j, i) /= weights_set[i];
+		}
+	}
+}
+
+void CKMeansBase::set_initial_centers(SGVector<float64_t> weights_set, 
+				SGVector<int32_t> cl_list, int32_t XSize)
+{
+	REQUIRE(m_mus_initial.matrix, 
+			"Initial cluster centers not supplied.\n");
+
+	/// set rhs to mus_start
+	CDenseFeatures<float64_t>* rhs_mus=new CDenseFeatures<float64_t>(0);
+	CFeatures* rhs_cache=distance->replace_rhs(rhs_mus);
+	rhs_mus->copy_feature_matrix(m_mus_initial);
+
+	SGVector<float64_t> dists=SGVector<float64_t>(k*XSize);
+	dists.zero();
+
+	for(int32_t idx=0;idx<XSize;idx++)
+	{
+		for(int32_t m=0;m<k;m++)
+			dists[k*idx+m] = distance->distance(idx,m);
```

Comment:
no need to loop, the distance class should be able to compute all distances itself (and in parallel)